feat: add organization-scoped permission checks to deployment settings

code-asher opened the pull request:
Part of https://github.com/coder/coder/issues/13915

You should be able to see the all the links on the settings page to which you have access, and those pages should react appropriately (like groups should only have the create group button if you have permission, etc).

This only affects multi-org.  There should be no change if not multi-org.

code-asher committed:
s/readAllUsers/viewAllUsers

Seems like all the other frontend variables use the `view` syntax.
Arguably we should use `read` to match the backend, but `view` does seem
more UI-like.

code-asher committed:
Check license for organizations

All the checks now require both the experiment and license.

I also renamed the variable canViewOrganizations everywhere for
consistency.

code-asher committed:
Allow any auditor to view the audit log

code-asher committed:
Scope create group button to organizations

You will only see the button if you can create a group in that org.

Now that the old site-wide group permission is unused, remove it.

code-asher committed:
Use fine-grained permissions on settings page

Since in addition to deployment settings this page now also includes
users, audit logs, groups, and orgs.

Since you might not be able to fetch deployment values, move all the
loaders to the individual pages instead of in the wrapping layout.

BrunoQuaresma commented on site/src/pages/DeploySettingsPage/DeploySettingsLayout.tsx:
```diff
@@ -52,19 +55,15 @@ const DeploySettingsLayoutInner: FC = () => {
-            ) : (
-              <Loader />
-            )}
+            <DeploySettingsContext.Provider
```
When checking if the data exists, we can ensure the children's pages can access the data directly without needing to handle the loading state. Removing this check requires each page to deal with the loading itself(adding an extra amount of work). Maybe I'm missing something but I don't see value in this change.

code-asher replied:
Not all users coming to this page can access the deployment settings.  Some of them can only access the organization information or only the audit log, so if we add loading here then those users can never see the page as they will only see a loader.  So we have to handle the loader on each individual page since they all have kinds of permissions.

I think maybe we could do this differently by not combining different things into the same page, but that would require a rework of the design.

code-asher replied:
An alternative would be to always run the query and just let it error for users without the right permissions.  We could then move the loader back here.  But we would still need to handle the error state on each page.  Or we could ignore the error state, since users should not be seeing links to those pages anyway since they lack permissions.

BrunoQuaresma replied:
> Not all users coming to this page can access the deployment settings. Some of them can only access the organization information or only the audit log or only the users etc, 

Makes sense. 👍

BrunoQuaresma commented on site/src/pages/ManagementSettingsPage/GroupsPage/GroupsPage.tsx:
```diff
@@ -3,53 +3,69 @@ import Button from "@mui/material/Button";
+    organization: string;
+  };
+  const groupsQuery = useQuery(
+    organizationName ? groups(organizationName) : { enabled: false },
```
TIL! I was doing:
```tsx
const groupsQuery = useQuery({
    ...groups(organizationName ?? ""),
    { enabled: !!organizationName }
})
```
But your way is better.

code-asher replied:
I think I stole this from Kayla actually :laughing:

BrunoQuaresma commented on site/src/pages/ManagementSettingsPage/OrganizationSettingsPage.tsx:
```diff
@@ -26,32 +29,88 @@ const OrganizationSettingsPage: FC = () => {
+    organization
+      ? organizationPermissions(organization.id)
+      : { enabled: false },
+  );
```
Maybe we could wrap this into a function like `getOrgPermissionsByName` since it is being used in a few places?

code-asher replied:
Yeah I could probably just bake in this behavior straight into `organizationPermissions` actually.

BrunoQuaresma reviewed (commented):
I left a few comments, and I see there are a few TODOs in the codebase. Do you plan to finish these TODOs during the current work or address them later? If later, would it make sense to transform the important ones into issues?

PS: I'm going to QA this at some point today.

BrunoQuaresma commented:
I QAed the changes and they look good, but I think it could be better if you have specific areas where I should look for issues. Since it is a "draft" PR I'm not approving it for now, but feel free to request another review at any time.

code-asher commented on site/src/pages/ManagementSettingsPage/GroupsPage/GroupsPage.tsx:
```diff
@@ -3,53 +3,69 @@ import Button from "@mui/material/Button";
+    organizationName ? groups(organizationName) : { enabled: false },
+  );
   const { organizations } = useOrganizationSettings();
+  // TODO: If we could query permissions based on the name then we would not
```
Yeah, I was not sure if it would be something others wanted but if so then definitely we should make an issue.

code-asher commented on site/src/pages/ManagementSettingsPage/GroupsPage/GroupsPage.tsx:
```diff
@@ -3,53 +3,69 @@ import Button from "@mui/material/Button";
+    if (defaultName) {
+      return <Navigate to={`/organizations/${defaultName}/groups`} replace />;
+    }
+    return <EmptyState message="No default organization found" />;
```
Code-wise yeah but in practice?  Probably not.  You would somehow need to not have a default org, maybe if there was some database corruption or something idk

code-asher replied:
Oh huh yeah that is a much better idea.  Will change!

code-asher commented:
@BrunoQuaresma Thank you for giving it an early look!  I still need to add some tests/Storybooks and then I think it will be good to go.  I will make some notes on the kinds of things that need to be QA'd.

As for the TODOs, I will check with Kira and make issues for the ones she thinks makes sense.

code-asher committed:
Dedupe disable org permission query

code-asher committed:
Show better message when unable to edit anything in an org

code-asher committed:
Fix edit org member check

Looks like the resource needs to be organization_member, not user.

Also we were not passing through the edit permission to the edit and
delete buttons.

code-asher committed:
Update org member tests

- Was not using the right layout.
- Add feature and permissions responses.
- Must use beforeEach, I guess routes get reset somewhere.
- A bit unrelated, but change the useQuery to match the format of the
  others (avoiding the spread).

code-asher committed:
Add stories for organization members page

Needed to break it out into a separate view to do this.

code-asher committed:
Add stories for multi-org sidebar

code-asher committed:
Remove multi-org check from management settings layout

We only use this layout when multi-org is enabled, so no need to run the
check a second time.

code-asher committed:
Check for license permission

An auditor for example appears to have the view deployment values
permission but cannot view licenses.

code-asher commented:
All right, I updated the tests and added stories but I had to break out into some views to do that.  The code is mostly unchanged but the diff will look a a bit large.

So generally here is what I have been testing:

- A site-wide admin should see and be able to edit all the deployment settings and organizations.
- A site-wide auditor should see the audit logs for the default org.  And the deployment settings apparently, but not the licenses or appearance.  They can also see the members and groups for the default orgs but not edit either.
- An org admin should only be able to edit the settings, members, and groups of orgs of which they are an admin, and should see the audit logs of their orgs as well.
- All of the above should be able to view users site-wide.  But only the admin can add new users.
- An org auditor should only be able to see the audit logs for their own orgs.  But no access to the members or groups of the org.
- Plain members will not see any deployment dropdown at all (just like before).
- If you have multi-org disabled or are not licensed for multi-org then everything should look and work exactly the way it was before.

NOTE: The audit log will not work right now for org admins/auditors, but they will once Steven's PR is merged so I will leave this in a draft until that is done then I will mark ready and request a review.  <3

code-asher commented:
Should be good to go now!

code-asher committed:
Add more stories for deployment dropdown

BrunoQuaresma commented:
- ✅  A site-wide admin should see and be able to edit all the deployment settings and organizations.
- ❌ A site-wide auditor should see the audit logs for the default org. And the deployment settings apparently, but not the licenses or appearance. They can also see the members and groups for the default orgs but not edit either.
- ⚠️ An org admin should only be able to edit the settings, members, and groups of orgs of which they are an admin, and should see the audit logs of their orgs as well.
- All of the above should be able to view users site-wide. But only the admin can add new users.
- An org auditor should only be able to see the audit logs for their own orgs. But no access to the members or groups of the org.
- ✅  Plain members will not see any deployment dropdown at all (just like before).
- ✅  If you have multi-org disabled or are not licensed for multi-org then everything should look and work exactly the way it was before.

1. As an auditor, I was able to see the licenses.
<img width="1512" alt="Screenshot 2024-08-05 at 10 24 55" src="https://github.com/user-attachments/assets/55e49a7f-8844-409c-b169-9
<img width="1055" alt="Screenshot 2024-08-05 at 10 26 45" src="https://github.com/user-attachments/assets/03a873c4-50ec-41b9-9861-c34fa58824da">
04f93fb4f5f">

2. As an auditor, I'm getting stuck on the audit logs page. I can't see the other org options.

https://github.com/user-attachments/assets/b3c4f58b-abd4-4d73-b933-6169b94fce92

3. As org admin, I was able to remove the owner from the members. I don't think admins should have the right to remove admins or any other higher user level.

code-asher commented:
Thank you for the thorough testing!!

>  As an auditor, I was able to see the licenses.

Do you know if you were on the latest commit for the PR?  I think I already fixed this since I was not able to reproduce.

> As an auditor, I'm getting stuck on the audit logs page. I can't see the other org options

Huh, weird I was not able to reproduce with a site-wide admin.  I will poke into it more.

> As org admin, I was able to remove the owner from the members. I don't think admins should have the right to remove admins or any other higher user level.

That does seem weird, let me check with the backend folks.

code-asher committed:
Throw error if no default org

code-asher committed:
Merge branch 'main' into asher/org-permissions

code-asher committed:
Use default org ID in old pages

The 0000 syntax does not work, it has to be the actual org ID.

Technically this is probably not necessary since it means only admins
can create groups which was already the case, but I think being explicit
about the org we are checking is a good idea, even if these pages will
eventually go away in favor of the multi-org ones.

code-asher committed:
Remove org permission-based redirects

I think this needs a proper design/decision.

code-asher commented:
I changed the automatic redirect for auditors, so now it just says "you have no permissions to edit" if there are no permissions.  I think we will probably want to consider a better design for this though.

code-asher committed:
Use old permission for old pages

Originally I thought I should update these to use the org-based
permissions query but I think maybe I should scope my changes to just
the multi-org settings page and leave the old pages alone for now.

code-asher committed:
Show ineditable form without org edit perm

I think the "you do not have permission to edit" message might make it
sound like you can do nothing at all with the org, but it might be that
you can edit members and groups, just not the org settings.

So, instead, show the org settings but disable the form if you cannot
edit.  Probably there is still a better design but maybe this is OK for
now.

code-asher commented:
Just FYI I reverted two changes to the old groups and users pages, I felt like they were out of scope for this PR and I did not mention them in the list of stuff to test.

I also changed the org settings page to show an ineditable form instead of "you cannot edit this org" but I will get a proper design for this page when a user has no permissions later.
